### ISSUE TITLE
Add a mechanism to group multiple layers sources with the same name but different version

### ISSUE THREAD

**Issue by Guts** [aeag/MenuFromProject-Qgis-Plugin#135]
## Use case

Group layers with the same base name but different versions into a single entry in the layer tree, and implement a contextual menu for version selection within the explorer.

## Expected behavior

On double-click or selecting the "Display Layer" option in the context menu, the application defaults to the first version and format found. Versions are sorted based on the order of layers in the source project. If only one version and format are defined, the "Versions" menu disappears from the context menu.

![image](https://github.com/user-attachments/assets/6c6d2141-dd3f-4cfd-a533-3d8b0c91fe0e)

## Implementation proposal

Some ways we've though about.

### A. New syntaxic tip

Use a specific character in layer names to extract the version information.

Example:

`communes$2023` and `communes$2024` → Displayed as "Communes" with a submenu "Versions" with "2023 and "2024"

### B. Rely on layer's metadata

Leverage a specific metadata keyword for versioning or a date(time) field.

Exemple:

Add a dedicated keyword like `version:2024` in the layer metadata.

![image](https://github.com/user-attachments/assets/14858e72-36f5-4de1-ad7d-38abd864385f)

### C. Use the [layer's auxiliary storage](https://docs.qgis.org/3.34/en/docs/user_manual/working_with_vector/vector_properties.html#auxiliary-storage-properties)

Use auxiliary storage associated with the layer to manage versioning.

### Additional context

This feature request is formulated from [Agences de l'Eau](https://www.lesagencesdeleau.fr/) and ported by [Oslandia](https://oslandia.com/).

**Comment by jmkerloch** [comment]
If we use the new syntaxic tip, should we keep the version in the loaded layer name ?

**Comment by xcaeag** [comment]
I'd tend to say yes, replacing the $ with a space?

Communes$2023  -> Communes 2023
Communes$v2023  -> Communes v2023

**Comment by naub1n** [comment]
> If we use the new syntaxic tip, should we keep the version in the loaded layer name ?

like @xcaeag , yes :wink:

**Comment by jmkerloch** [comment]
I suggest that we also group multiple layers only inside a group. With this approach we can still have same layer name in a different group.

For example with this project:

![Image](https://github.com/user-attachments/assets/5e9e2b46-cee4-40cd-91ff-0b88853ef15f)

This will be like this in menu

https://github.com/user-attachments/assets/f06a53af-0f1e-4676-8784-db50bde8327f

And in browser:

https://github.com/user-attachments/assets/5c072a65-edd9-4b2f-b0dc-5b709d000e61

**Comment by jmkerloch** [comment]
After some tests, I think it will be better if all versions are in a `Versions`menu.

With this approach, all layer with version or format will have the same number of sub menu:

- `Load layer`
- `Versions`

**Comment by xcaeag** [comment]
I don't have a clear-cut opinion, you have to keep things simple and practical. 
question: does double-clicking still open the layer without having to open the context menu ? Is this contextual menu necessary in the absence of different versions?

**Comment by jmkerloch** [comment]
> I don't have a clear-cut opinion, you have to keep things simple and practical. question: does double-clicking still open the layer without having to open the context menu ? Is this contextual menu necessary in the absence of different versions?

If there is no multiple version then the menu `Versions` is not available.

On the browser, double click open the first available layer if there are multiple versions or format. 

On the menu there is no double click option, we need to go to `Load layer`

**Comment by naub1n** [comment]
> After some tests, I think it will be better if all versions are in a `Versions`menu.
> 
> With this approach, all layer with version or format will have the same number of sub menu:
> 
>     * `Load layer`
> 
>     * `Versions`

Ok, it's like in the initial "Expected behavior"

**Comment by jmkerloch** [comment]
How should we support layer that have same name but different geometry type:

![Image](https://github.com/user-attachments/assets/19a3ab38-0d84-4f59-bb0c-7fbf7614b150)

For `sample` layer  should we have only one layer in the menu ?

**Comment by jmkerloch** [comment]
For the format definition we are currently getting the information from the provider defined in the layer.

To be able to distinguish WMS and WMTS, this information won't be enough. WMS and WMTS layer use the same provider.

I think we have no other choice than checking the layer metadata to get the format for LMFP.

I suggest a variable named `LMFP_FORMAT`.

This will be the first place checked, otherwise we will keep the provider defined in the layer.

**Comment by xcaeag** [comment]
Yes, that's fine, the use cases will be rare, and users won't often be able to tell the difference between WMS and WMST, so...
Another solution would have been to add a $, allowing any word to be placed?

- mnt$2021$wms
- mnt$2021$wmst
- mnt$2021$web

**Comment by jmkerloch** [comment]
> Yes, that's fine, the use cases will be rare, and users won't often be able to tell the difference between WMS and WMST, so... Another solution would have been to add a $, allowing any word to be placed?
> 
>     * mnt$2021$wms
> 
>     * mnt$2021$wmst
> 
>     * mnt$2021$web

I'm not sure if new syntaxic tip are welcome. I will first add support for metadata.

**Comment by naub1n** [comment]
> How should we support layer that have same name but different geometry type:
> 
> ![Image](https://github.com/user-attachments/assets/19a3ab38-0d84-4f59-bb0c-7fbf7614b150)
> 
> For `sample` layer should we have only one layer in the menu ?

Yes, layer with the same name should have only one menu. We don't mix layers with different geometry. 

> For the format definition we are currently getting the information from the provider defined in the layer.
> 
> To be able to distinguish WMS and WMTS, this information won't be enough. WMS and WMTS layer use the same provider.
> 
> I think we have no other choice than checking the layer metadata to get the format for LMFP.
> 
> I suggest a variable named `LMFP_FORMAT`.
> 
> This will be the first place checked, otherwise we will keep the provider defined in the layer.

I suggest to talk about this feature in a short meeting. 
I want to know if it's a good idea to find good provider with internal LMFP process (like find `GetTile` value in `GetCapabilities` request).
@xcaeag is also a good idea. It's the easiest for adminstrators.

**Comment by naub1n** [comment]
Summary of the meeting:
- Hight priority on automatic provider matching.
- Metadata variable is already implemented and can be used as a last resort.

LMFP use:
1. LMFP_FORMAT if exists.
2. Otherwise, it use internal process to find good provider
3. Otherwise, it use QGIS Provider

If some providers are not indentify by internal process, we can make a suggestion with standard information to use, in another plugin update.

**Comment by jmkerloch** [comment]
@naub1n for the internal process to distinguis WMS and WMTS, I have some concern about the use of GetCapabilities, this could cause some delay for project parsing because of the request.

I suggest we look in the source of the WMS and look for specific value. I looked at QGIS code and this is what's done to define if a layer is tiled.

For example:

- `&type=xyz`  : we have XYZ Tiles
- `&type=mbtiles`  : we have Mb Tiles
- `&type=wmst`:  we have WMST

To be more generic, if `type` is in source we use it to define format

If no type we could also check `tileMatrixSet` and `tileDimensions` definition. If a value is set we will use WMTS format.

**Comment by naub1n** [comment]
> [@naub1n](https://github.com/naub1n) for the internal process to distinguis WMS and WMTS, I have some concern about the use of GetCapabilities, this could cause some delay for project parsing because of the request.
> 
> I suggest we look in the source of the WMS and look for specific value. I looked at QGIS code and this is what's done to define if a layer is tiled.
> 
> For example:
> 
>     * `&type=xyz`  : we have XYZ Tiles
> 
>     * `&type=mbtiles`  : we have Mb Tiles
> 
>     * `&type=wmst`:  we have WMST
> 
> 
> To be more generic, if `type` is in source we use it to define format
> 
> If no type we could also check `tileMatrixSet` and `tileDimensions` definition. If a value is set we will use WMTS format.

Ok, perfect ! `tileMatrixSet` is a required parameter for WMTS.
